File: Problems 0000-0099/16. 3Sum Closest/solution.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        n = len(nums)
        nums.sort()
        closest = float('inf')
        for i in range(0, n - 2):
            # skip repeat
            if i > 0 and nums[i] == nums[i-1]:
                continue
            
            # optimization #1:
            maxSum = nums[i] + nums[-2] + nums[-1]
            if maxSum < target:
                # then no need to go any further
                if abs(target - maxSum) < abs(target - closest):
                    closest = maxSum
                continue
            
            # optimization #2:
            minSum = nums[i] + nums[i+1] + nums[i+2]
            if minSum > target:
                if abs(target - minSum) < abs(target - closest):
                    closest = minSum
                break
            
            l, r = i + 1, n - 1
            while l < r:
                sumVal = nums[i] + nums[l] + nums[r]
                if abs(target - sumVal) < abs(target - closest):
                    closest = sumVal
                if sumVal == target:
                    return target
                elif sumVal < target:
                    l += 1
                else:
                    r -= 1
        return closest
        

File: Problems 0000-0099/16. 3Sum Closest/solution.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        if (nums.size() < 3) return 0;
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int res = nums[0] + nums[1] + nums[2];
        for (int i = 0; i < n - 2; i++) {
            if (i > 0 && nums[i] == nums[i-1]) continue;

            // this is the closest to target in this iteration
            // no need to go further and try every combo
            int maxSum = nums[i] + nums[n-2] + nums[n-1];
            if (maxSum < target) {
                if (abs(target-maxSum) < abs(target - res)) {
                    res = maxSum;
                }
                continue;
            }

            // any further combo will only get bigger and further
            // away from target, so break
            int minSum = nums[i] + nums[i+1] + nums[i+2];
            if (minSum > target) {
                if (abs(target-minSum) < abs(target-res)) {
                    res = minSum;
                }
                break;
            }

            int l = i+1;
            int r = n-1;
            while (l < r) {
                int total = nums[i] + nums[l] + nums[r];
                if (abs(target - total) < abs(target - res)) {
                    res = total;
                }
                if (total == target) {
                    return total;
                } else if (total < target) {
                    l++;
                } else {
                    r--;
                }
            }
        }
        return res;
    }
};